File: app/src/main/java/com/osrsd/cache/provider/ParamProvider.java
package com.osrsd.cache.provider;

import com.osrsd.cache.def.ParamDefinition;
import com.osrsd.cache.util.ByteBufferExt;
import com.osrsd.spring.domain.Definition;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;

@Slf4j
public final class ParamProvider {

    public static Definition decode(ByteBuffer buffer, ParamDefinition definition) {
        do {
            int opcode = buffer.get() & 0xff;
            if (opcode == 0) {
                break;
            }

            if (opcode == 1) {
                definition.setType((char) buffer.get());
            } else if (opcode == 2) {
                definition.setDefaultInt(buffer.getInt());
            } else if (opcode == 4) {
                definition.setMembers(false);
            } else if (opcode == 5) {
                definition.setDefaultString(ByteBufferExt.INSTANCE.getString(buffer));
            } else {
                log.warn(String.format("Unhandled definition opcode with id: %s.", opcode));
            }
        } while (true);
        return definition;
    }

}

File: app/src/main/kotlin/com/osedit/cache/provider/ParamProvider.kt
package com.osedit.cache.provider

import com.osedit.cache.util.ByteBufferExt
import com.osedit.spring.domain.Definition
import com.osedit.spring.domain.ParamDefinition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer

object ParamProvider : Provider<ParamDefinition> {

    private val log: Logger = LoggerFactory.getLogger(ParamProvider::class.java)

   override fun decode(buffer: ByteBuffer, definition: ParamDefinition): Definition {
        do {
            when (val opcode: Int = buffer.get().toInt() and 0xff) {
                1 -> definition.type = (buffer.get().toInt() and 0xff).toChar()
                2 -> definition.defaultInt = buffer.int
                4 -> definition.isMembers = false
                5 -> definition.defaultString = ByteBufferExt.getString(buffer)
                0 -> break
                else -> log.warn(String.format("Unhandled definition opcode with id: %s.", opcode))
            }
        } while (true)
        return definition
    }

}
